### PR TITLE
Support function authentication with OpenFaaS IAM

### PR DESCRIPTION
**Pull request by welteki**
## Description

If JWT authentication is enabled by setting the env variable `jwt_auth: true` in the `stack.yml` file the watchdog requires an OpenFaaS function token for request authentication.

Authentication tokens can be optioned through OpenFaaS IAM.

The watchdog verifies the token is valid and checks of an actor is authorized to invoke the function by looking at the `function` claim in the token and validating the permissions.

## Motivation and Context

- [ ] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

Support IAM authentication for functions.

## How Has This Been Tested?

- Unit tests were added.
- Changes have been tested by running the watchdog locally and invoking it with and without a function token.
During these test the `OPENFAAS_NAME` and `OPENFAAS_NAMESPACE` env variable were changed to simulate different functions in multiple namespaces.

    ```sh
    fprocess="env" \
    mode="serializing" \
    jwt_auth=true \
    jwt_auth_local=true \
    jwt_auth_debug=true \
    OPENFAAS_NAMESPACE=staging \
    OPENFAAS_NAME=env \
    port=8083 ./bin/fwatchdog-amd64
    ```
- Changes have been tested E2E in cluster with the updated oidc-provider.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: executor/jwt_authenticator.go [hunk 1]
```diff
@@ -179,3 +196,76 @@ type OpenIDConfiguration struct {
...
+func getFnName() (string, error) {
+	name, ok := os.LookupEnv("OPENFAAS_NAME")
+	if !ok || len(name) == 0 {
+		return "", fmt.Errorf("env variable 'OPENFAAS_NAME' not set")
+	}
+
+	return name, nil
+}
+
+func getFnNamespace() (string, error) {
```
Comment (alexellis) [inline executor/jwt_authenticator.go:209]: How would we run this with go run ?

Comment (welteki) [reply executor/jwt_authenticator.go:209]: I made the path used to lookup the namespace configurable by setting the env variable `namespace_mount_path`. This allows you to set any namespace name when running a function locally with Docker or `faas-cli local-run`

e.g for local testing the watchdog with jwt_auth:

```sh
fprocess="env" \
mode="serializing" \
jwt_auth=true \
jwt_auth_local=true \
namespace_mount_path=./secrets/namespace \
OPENFAAS_NAME=env \
./bin/fwatchdog-amd64
```

Comment (alexellis) [reply executor/jwt_authenticator.go:209]: I think I'd prefer it to be `OPENFAAS_NAMESPACE` as a priority order.

Comment (welteki) [reply executor/jwt_authenticator.go]: This has been changed. The env variable `OPENFAAS_NAMESPACE` is now used to lookup the namespace if it is not set the watchdog will try to read the namespace from `/var/run/secrets/kubernetes.io/serviceaccount/namespace`.

File: main.go [hunk 2]
```diff
@@ -474,3 +495,26 @@ func (nc *WriterCounter) Write(p []byte) (int, error) {
 	nc.bytes += int64(n)
 	return n, err
 }
+
+func getFnName() (string, error) {
+	name, ok := os.LookupEnv("OPENFAAS_NAME")
```
Comment (alexellis) [inline main.go:500]: Should `faas-cli local-run` inject this variable then? I don't think it does presently.

cc @rgee0

**Review by alexellis — APPROVED** [review]
Approved

**Pull request merged**